Cabal/Distribution/Package.hs

```diff
@@ -187,24 +189,76 @@ instance NFData ComponentId where
 
 -- | Returns library name prefixed with HS, suitable for filenames
 getHSLibraryName :: UnitId -> String
-getHSLibraryName (SimpleUnitId (ComponentId s)) = "HS" ++ s
+getHSLibraryName uid = "HS" ++ display uid
 
--- | For now, there is no distinction between component IDs
--- and unit IDs in Cabal.
-newtype UnitId = SimpleUnitId ComponentId
-    deriving (Generic, Read, Show, Eq, Ord, Typeable, Data, Binary, Text, NFData)
+-- | A unit identifier identifies a (possibly instantiated)
+-- package/component that can be installed the installed package
+-- database.  There are several types of components that can be
+-- installed:
+--
+--  * A traditional library with no holes, so that 'unitIdHash'
+--    is @Nothing@.  In the absence of Backpack, 'UnitId'
+--    is the same as a 'ComponentId'.
+--
+--  * An indefinite, Backpack library with holes.  In this case,
+--    'unitIdHash' is still @Nothing@, but in the install,
+--    there are only interfaces, no compiled objects.
+--
+--  * An instantiated Backpack library with all the holes
+--    filled in.  'unitIdHash' is a @Just@ a hash of the
+--    instantiating mapping.
+--
+-- A unit is a component plus the additional information on how the
+-- holes are filled in. Thus there is a one to many relationship: for a
+-- particular component there are many different ways of filling in the
+-- holes, and each different combination is a unit (and has a separate
+-- 'UnitId').
+--
+-- Since the mapping is not actually recorded in a 'UnitId',
+-- you can't actually substitute over them.  Look at 'IndefUnitId'
+-- for a variant that can explicitly record module variables,
+-- as well as "Distribution.Backpack.FullUnitId" for a mechanism for
+-- expanding an instantiated 'UnitId' to retrieve its mapping.
+--
+data UnitId
+    = UnitId {
+        unitIdComponentId :: ComponentId,
+        unitIdHash        :: Maybe String
```

Comment:
We can let @hvr to come along and turn this into an opaque newtype represented by a ShortByteString :-)